src/query/set_query.rs

```diff
@@ -0,0 +1,198 @@
+use std::collections::BTreeSet;
+
+use tantivy_fst::raw::CompiledAddr;
+use tantivy_fst::{Automaton, Map};
+
+use crate::query::{AutomatonWeight, Query, Weight};
+use crate::schema::Field;
+use crate::{Searcher, Term};
+
+/// A Term Set Query matches all of the documents containing any of the Term provided
+///
+/// Terms not using the right Field are discared.
+#[derive(Debug, Clone)]
+pub struct TermSetQuery {
+    field: Field,
+    terms: BTreeSet<Term>,
+}
+
+impl TermSetQuery {
+    /// Create a Term Set Query
+    pub fn new(field: Field, terms: BTreeSet<Term>) -> Self {
```

Comment:
`BTreeSet` is a good data structure for a set that is continuously changing while always staying sorted and using the sorting to speed element access. In this case, it seems that the set is accessed only once to produce a weight.

Maybe it would be nicer to simply require `terms: T where T: IntoIterator<Item=Term>`, collect this into a `Vec` (where `terms.into_iter().collect::<Vec<_>>()` would not allocate if the iterator is create from a `Vec`) and sort and deduplicate this `Vec` once in this constructor?

I think this could yield nicer API and simpler and hence faster code, but then again it could be insignificant and thereby not worth it especially if one expects the terms to be already presented as a `BTreeSet`.

Reply:
I like the BTreeSet personally.

Reply:
> I like the BTreeSet personally.

I would not say that I dislike it (or any other data structure for that matter), just that it brings more to the table than is required here. I do think it is arguably more complex than say `Vec` which is why I tried to suggest the simplest possible data structure for this particular task.

Reply:
What should I do then? Personally I prefer using a `BTreeSet`, but I understand an `IntoIterator` is easier to provide in general

Reply:
As a potential user of this query*, if the type was a `BTreeSet` we would have to create one just for this purpose, and the query just needs immutable sorted & deduped data for which `Vec` suffices. I would vote for `IntoIterator`.


* -- currently doing union boolean query over tens of IDs -- looks like this should be much more efficient!

Reply:
IntoIterator it is then...

and then within the function:
`IntoIterator -> HashSet -> Vec -> Sort.`